src/merge_iterator.rs

```diff
@@ -89,19 +127,20 @@ where
 
 impl<T1: KeyValueIterator, T2: KeyValueIterator> KeyValueIterator for TwoMergeIterator<T1, T2> {
     async fn next_entry(&mut self) -> Result<Option<RowEntry>, SlateDBError> {
-        if let Some(next1) = self.iterator1.1.as_ref() {
-            if let Some(next2) = self.iterator2.1.as_ref() {
-                if next1.key <= next2.key {
-                    if next1.key == next2.key {
-                        self.advance2().await?;
-                    }
-                    return self.advance1().await;
-                }
-                return self.advance2().await;
+        let mut current_kv = match self.advance_inner().await? {
+            Some(kv) => kv,
+            None => return Ok(None),
+        };
+        while let Some(peeked_kv) = self.peek_inner() {
```

Comment:
One thing I'm still not sure about is does it make sense to filter out repeated keys from this iterator or from the memtable iterators? I think one aspect is how we implement snapshots, and whether or not we anchor a snapshot to a checkpoint. If each snapshot has a checkpoint, then none of the SSTs actually need to store multiple versions of a key unless the rows are merge entries. Only memtables need to store multiple versions of rows. So we'd only need this logic for the memtables in that case. On the other hand, it doesn't really do any harm to put it here (other than to make the impl more complex).

Maybe the best thing to do would be to let the creator of the iterator configure how they want to handle multiple instances of the same key by passing in some `fn`? And then the caller can configure the iterators as appropriate. Is this along the lines of what you were thinking for merge @dmvk ?

Reply:
as far as i know, the lsm engines mostly have multiple versions in the sst files. i tend to follow their approaches at the moment.

this approach is well understood to have snapshot implemented with solid proof. i'm also interested in your suggestion about do not have multiple versions in SST files, i'd like to have this discussed in an RFC to discuss the details on cleanups of the legacy snapshot data. (my intuition tells me it'll have fewer risks to follow the existing approaches though 🤔)

> If each snapshot has a checkpoint, then none of the SSTs actually need to store multiple versions of a key unless the rows are merge entries

creating a checkpoint for each snapshot is not in consideration, it'll hurt performance a lot, as snapshot will be created on every range, it should be a lightweight operation to create a snapshot. am i understand wrong about your proposal?

Reply:
I suppose the reason why they do that is to conserve space, but we don't have to be as concerned about that because we store everything in the object store which has no hard limits and is relatively cheap. I suppose they would also have to include some sort of GC, which slatedb already has as part of it's basic design. One benefit of using checkpoints is that it can make the compaction logic and sst iteration simpler - the compactor/iterators don't have to worry about multiple versions per key and about tracking a safe sequence number up to which it can discard or merge updated rows.

> please correct me if I'm misunderstanding your proposal. but creating a checkpoint for each snapshot is not under consideration, as it would hurt performance a lot. snapshots are intended to be created for every range scan, and this operation should be an read-only lightweight operation.

No, we wouldn't/shouldn't create a checkpoint per snapshot. I was thinking we could instead tie snapshots to the checkpoint that the writer would already maintain once the checkpoints RFC is completed (see https://github.com/slatedb/slatedb/blob/main/rfcs/0004-checkpoints.md#writers). So each new snapshot adds a reference to the writer's current checkpoint, and then the checkpoint is deleted when there are no remaining references.

Reply:
> 🤔 if we decide to adopt the checkpoint-per-snapshot approach later, we can revisit this code path in another pr, wdyt?

Yeah I'm fine with this. My main suggestion was to see if we could factor out the merge logic into something that could be passed into the iterator. E.g. have the iterator take something like a `merger: Fn(rows: &[RowEntry]) -> RowEntry` that owns how the rows are combined rather than bake it into the iterators, mostly to keep this logic decoupled from the main iterator implementation so it doesn't become a headache to change later if we want to. I'd guess we'll want something like this anyway for the merge operator anway.

Reply:
> > 🤔 if we decide to adopt the checkpoint-per-snapshot approach later, we can revisit this code path in another pr, wdyt?
> 
> Yeah I'm fine with this. My main suggestion was to see if we could factor out the merge logic into something that could be passed into the iterator. E.g. have the iterator take something like a `merger: Fn(rows: &[RowEntry]) -> RowEntry` that owns how the rows are combined rather than bake it into the iterators, mostly to keep this logic decoupled from the main iterator implementation so it doesn't become a headache to change later if we want to. I'd guess we'll want something like this anyway for the merge operator anway.

I tend to think `seqnum` is a fundamental primitive in the SST/Memtable encoding, and we should keep it visible when designing iterators as part of the overall design, because the design of iterators and the design of SSTs are tightly integrated. it's not a general iterator, but an iterator for `RowEntry` only. this merge operation only exists when there are multiple versions of a row, and seqnum is the only indicator to pick value from multiple versions.

I think we can explore whether there are other possibilities that would require a similar deduplication operation and could leverage this abstraction capability. Personally, I prefer to introduce abstractions when encountering specific problems, as this often helps us discern which abstractions are more appropriate.

Reply:
> No, we wouldn't/shouldn't create a checkpoint per snapshot. I was thinking we could instead tie snapshots to the checkpoint that the writer would already maintain once the checkpoints RFC is completed (see https://github.com/slatedb/slatedb/blob/main/rfcs/0004-checkpoints.md#writers). So each new snapshot adds a reference to the writer's current checkpoint, and then the checkpoint is deleted when there are no remaining references.

I don't fully understand why there can be no multiple versions in the SST files in this design, can you help give an explanation for this? 😲

after a snapshot is created, the memtable is still increasing, right? when l0 reached its limit, it should create an l0 sst file with multiple versions in it. and when l0 reached its limit, l1 is created. the multi version data should still kept in l0 and l1 files. do i understand wrong about this?

Reply:
Let's suppose the writer's current checkpoint C points to a manifest with something like `l0: [l0-0, l0-1, l0-3, l0-4], srs: [sr1, sr2]`, and new data are being written to memtable. If I create a new snapshot tied to C, then all the reads that the snapshot does will be issued against the SSTs from C. So if a compaction runs (let's say compacting all the L0 ssts in to sr1), it can freely remove keys in the resulting sorted run without worrying about disrupting the snapshot's reads.

We do still need to be able to deal with multiple versions in the memtable though. And eventually that memtable will need to be flushed. We would either need to have the snapshot keep a reference to the unflushed memtable so it can read from it (wastes memory), or we need to support multiple versions of keys in the L0 SSTs (but not in L1+)